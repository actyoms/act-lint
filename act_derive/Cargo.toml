[package]
name = "act_derive"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[[test]]
name = "ui"
path = "tests/tests.rs"

[dependencies]
attribute-derive = { version = "0.6.1", features = ["syn-full"] }
proc-macro2 = "1.0.66"
quote = "1.0.32"
serde = { version = "1.0.180", features = ["derive"] }
syn = { version = "2.0.28", features = ["full", "extra-traits"] }
trybuild = { version = "1.0.82", features = ["diff"] }
act_trait = { path = "../act_trait" }
