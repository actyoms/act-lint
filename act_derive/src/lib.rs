extern crate act_trait;
extern crate serde;

use proc_macro::TokenStream;

use attribute_derive::Attribute;
use proc_macro2::Ident;
use quote::quote;
use syn::{Data, DeriveInput, Generics, Result, Visibility};

mod expecting;

#[derive(Attribute, Debug)]
#[attribute(ident = act)]
struct ActAttribute {
    #[attribute(conflicts = [visit])]
    visitor: Option<String>,
    #[attribute(conflicts = [visitor])]
    visit: Option<Vec<String>>,
    expecting: Option<String>,
}

impl ActAttribute {
    fn is_valid_visit(v: &str) -> bool {
        matches!(v, "str" | "map")
    }
}

struct Input {
    name: Ident,
    attrs: ActAttribute,
    #[allow(dead_code)]
    vis: Visibility,
    #[allow(dead_code)]
    generics: Generics,
    #[allow(dead_code)]
    data: Data,
}

#[proc_macro_derive(Deserialize, attributes(act))]
pub fn deserialize_derive(tokens: TokenStream) -> TokenStream {
    all_attrs(tokens.clone())
        .map(|input| impl_deserialize(&input))
        .unwrap_or_else(syn::Error::into_compile_error)
        .into()
}

fn all_attrs(input: TokenStream) -> Result<Input> {
    let DeriveInput {
        ident,
        attrs: atts,
        vis,
        generics,
        data,
    } = syn::parse(input)?;
    ActAttribute::from_attributes(&atts)
        .and_then(|a| {
            if a.visit.is_none() && a.visitor.is_none() {
                Err(syn::Error::new_spanned(
                    ident.clone(),
                    "either `visit` or `visitor` must be specified in #[act(...)] attribute",
                ))
            } else {
                if let Some(visit) = &a.visit {
                    let err = visit
                        .iter()
                        .filter_map(|v| {
                            if !ActAttribute::is_valid_visit(v) {
                                Some(v.as_str().to_string())
                            } else {
                                None
                            }
                        })
                        .reduce(|a, b| format!("{}, {}", a, b))
                        .map(|v| {
                            syn::Error::new_spanned(
                                ident.clone(),
                                format!("invalid `visit` value(s): {} expected `str` or `map`", v),
                            )
                        });
                    if let Some(err) = err {
                        return Err(err);
                    }
                }
                Ok(a)
            }
        })
        .map(|attrs| Input {
            name: ident.clone(),
            attrs,
            vis,
            generics,
            data,
        })
}

fn impl_deserialize(input: &Input) -> proc_macro2::TokenStream {
    let attr = &input.attrs;
    let name = &input.name;
    let mut visitor = None;
    let visitor_x_impl = if attr.visitor.is_some() {
        quote! {}
    } else {
        visitor = Some(quote! { ActVisitor });
        visitor_token_stream(name, attr)
    };
    let expecting = expecting::impl_expecting(input);
    let gen = quote! {
        #expecting

        #visitor_x_impl

        impl<'de> serde::Deserialize<'de> for #name {
            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
            where
                D: serde::Deserializer<'de>,
            {
                deserializer.deserialize_any(#visitor)
            }
        }
    };
    // if let Some(expecting) = attr.expecting {
    //     expecting.into()
    // }
    gen
}

fn visitor_token_stream(name: &Ident, attr: &ActAttribute) -> proc_macro2::TokenStream {
    let mut need_str = false;
    let mut need_map = false;
    if let Some(visit) = &attr.visit {
        need_str = visit.contains(&"str".to_string());
        need_map = visit.contains(&"map".to_string());
    }
    let visit_str = visit_str_wrapper(name, need_str);
    let visit_map = visit_map_wrapper(name, need_map);
    quote! {
        struct ActVisitor;

        impl<'de> serde::de::Visitor<'de> for ActVisitor {
            type Value = #name;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                fn use_trait<T>(_method: fn(&mut std::fmt::Formatter) -> std::fmt::Result) {}
                use_trait::<#name>(<#name as act_trait::Expecting>::expecting);
                #name::expecting(formatter)
            }

            #visit_str

            #visit_map
        }
    }
}

fn visit_str_wrapper(name: &Ident, need: bool) -> proc_macro2::TokenStream {
    if need {
        quote! {
            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                <#name as std::str::FromStr>::from_str(v)
                    .map_err(|_| E::invalid_value(serde::de::Unexpected::Str(v), &self))
            }
        }
    } else {
        quote! {}
    }
}

fn visit_map_wrapper(name: &Ident, need: bool) -> proc_macro2::TokenStream {
    if need {
        quote! {
            fn visit_map<A>(self, map: A) -> Result<Self::Value, A::Error>
            where
                A: serde::de::MapAccess<'de>,
            {
                fn use_trait(_method: fn() -> bool) {}
                use_trait(<#name as act_trait::VisitMap<#name>>::is_map_visitor);
                #name::visit_map(map)
            }
        }
    } else {
        quote! {}
    }
}
