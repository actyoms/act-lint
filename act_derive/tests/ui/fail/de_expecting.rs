use act_derive::Deserialize;
use std::str::FromStr;

#[derive(Deserialize)]
#[act(visit = ["str"])]
struct MissingExpecting;

impl FromStr for MissingExpecting {
    type Err = std::io::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        todo!()
    }
}

fn main() {}
