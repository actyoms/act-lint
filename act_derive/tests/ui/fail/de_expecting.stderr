error[E0277]: the trait bound `MissingExpecting: Expecting` is not satisfied
 --> tests/ui/fail/de_expecting.rs:4:10
  |
4 | #[derive(Deserialize)]
  |          ^^^^^^^^^^^ the trait `Expecting` is not implemented for `MissingExpecting`
  |
  = note: this error originates in the derive macro `Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no function or associated item named `expecting` found for struct `MissingExpecting` in the current scope
 --> tests/ui/fail/de_expecting.rs:4:10
  |
4 | #[derive(Deserialize)]
  |          ^^^^^^^^^^^ function or associated item not found in `MissingExpecting`
5 | #[act(visit = ["str"])]
6 | struct MissingExpecting;
  | ----------------------- function or associated item `expecting` not found for this struct
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following traits define an item `expecting`, perhaps you need to implement one of them:
          candidate #1: `Visitor`
          candidate #2: `Expecting`
  = note: this error originates in the derive macro `Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)
